### PR TITLE
Adding Docker/docker-compose files

### PR DESCRIPTION
**Pull request by stakauskas**
Ability to build and run bitcoin from Docker. Builds docker image with bitcoin and DB4 4.8.30.NC, starts with docker-compose. Easy to modify startup parameters or add configuration files and scale containers. Image uses CentOS and gosu to start bitcoin as bitcoin user (PID 1).

### CODE REVIEW

**Comment by fanquake** [comment]
#10903 was the last time this was suggested. The conclusion then was that this isn't necessarily something we want to maintain in this repository, and would probably be better suited to an external repo.

See:
https://github.com/Sjors/docker-bitcoin-core
https://github.com/NicolasDorier/docker-bitcoin
https://github.com/jamesob/docker-bitcoind

for some examples of such repositories.

@stakauskas What's your motivation for adding the Dockerfiles here?

**Comment by MarcoFalke** [comment]
Agree with @fanquake

**Comment by promag** [comment]
NACK for the same reasons.

**Comment by stakauskas** [comment]
It was first created as an addition to ElementsProject in - https://github.com/ElementsProject/elements/pull/372 , thus appeared here.

Though having Docker images could be beneficial for community wanting to run software in containers and one image would target Windows/Mac/Linux instead of having separate packages for each of them.

**Comment by NicolasDorier** [comment]
I think maintaining several `Dockerfile` like this as one big advantage: We can use those as part of the CI to be sure bitcoind build correctly on the different OS which should be supported by Bitcoin, as well as running the tests on different distro.

@sipsorcery for example maintain a dockerfile for building on Windows via msvc, and configured CI to run it. This way he can warn you when something break Windows msvc build. 
Even if you are not fan of windows, I think you will see the advantage to have it for your favorite linux distro instead of relying on outdated documentation and untested dev environment.

It is also way more useful than maintaining a documentation, as anybody can reproduce it just by reading the dockerfile.

It also has the advantage that people can easily distribute their own bitcoin core image with their favorite freshly backed commits.

My repo https://github.com/NicolasDorier/docker-bitcoin is good if you want just an official release, but I totally see the advantage of providing dockerfiles.

It is big ACK for me.

**Comment by promag** [comment]
I consider these different:
 - Dockerfile(s) to build the *official* image - usually in a separate repository with multiple versions/variants - and usually download the office source code;
 - docker-compose to launch 3rd party services to aid development - not relevant here I think.
 - docker-compose + multiple Dockerfile to build from local source in different *flavours*  - usually to aid development or CI.

**Comment by promag** [comment]
> It also has the advantage that people can easily distribute their own bitcoin core image with their favorite freshly backed commits.

This is also possible with external Dockerfile.

**Comment by stakauskas** [comment]
I'd compare Dockerfile to RPM spec and docker-compose as a way to manage your containers, e.g systemctl/systemd for services, just portable across different OS versions. While you can run any docker image with 'docker run', it's much easier to start with docker-compose, which allows easier and faster configuration for running containers. In this case testnet/mainnet nodes can be started within one go, scale if needed. Network isolation/ports management provided out of the box, if you wanted 3 nodes, you do not need to change configuration for port mapping, as compose will assign random host -> same port in container, etc...

**Comment by MarcoFalke** [comment]
I don't see what advantage a dockerfile gives us to our existing processes. Note that it is already possible to build any commit from source deterministically by using it's gitian descriptor. Using a dockerfile to create a docker image as opposed to using gitian to create a binary wouldn't be deterministic any more.

Also note that the actual build step is just a single line `./configure && make`. Installing the dependencies is what differs per dist and usually the cause of issues. Instead of having a Dockerfile for each distribution, it seems easier to me to just list the dependencies in a readme (and maybe a one-liner for common distributions to install all dependencies)

Note that on travis we already use docker, but I don't see how a Dockerfile would simplify anything there.

If there is a valid use case for a Dockerfile that makes our process easier, I am all in, otherwise I think this will lead to a maintenance burden and nit-picking, because everyone has their own use case and wants the Dockerfile to match that use case.

**Comment by MarcoFalke** [comment]
> RPM spec

Sorry for ot, but we had someone contribute an rpm spec file and then it would just sit there unmaintained and now it is unclear what to do with it: #13137

I think we want to avoid doing the same with a Dockerfile.

**Comment by stakauskas** [comment]
> Installing the dependencies is what differs per dist and usually the cause of issues. Instead of having a Dockerfile for each distribution, it seems easier to me to just list the dependencies in a readme (and maybe a one-liner for common distributions to install all dependencies)

There's no need to have Dockerfile per each distribution, though it's possible and not difficult to maintain. After choosing base image, the whole lifecycle of running(also removing/updating/etc) software is as easy as choosing image:tag, bitcoin:version, instead of downloading package for Windows/Linux/Mac and figuring out deps individually.

**Comment by skwp** [comment]
As someone new to bitcoin compilation, I found getting a gitian build up and running quite complicated, especially on an OSX machine trying to use virtualization. 

I get the motivations around providing Dockerfiles in external projects but that means there is no (obvious) "source of truth" for getting started. If we want more devs poking at bitcoin I think it would be good that there was an official Dockerfile here that just gave you a build that you could play with locally (the intention being development, not production). That way you could test your changes easily. I was about to submit one but then saw this effort and decided to pile on my comment. 

I greatly support making it easier for new devs to get a local image running and docker provides exactly that.

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,102 @@
+FROM centos:centos7.4.1708
+
+ENV GOSU_VERSION 1.10
+ENV GOSU_ARCH amd64
+ENV GOSU_URL https://github.com/tianon/gosu/releases/download
+ENV GOSU_APP ${GOSU_URL}/${GOSU_VERSION}/gosu-${GOSU_ARCH}
+ENV GOSU_ASC ${GOSU_URL}/${GOSU_VERSION}/gosu-${GOSU_ARCH}.asc
+ENV DB4_VERSION 4.8.30.NC
+ENV DB4_HASH 12edc0df75bf9abd7f82f821795bcee50f42cb2e5f76a6a281b85732798364ef
+ENV DB4_URL http://download.oracle.com/berkeley-db/db-${DB4_VERSION}.tar.gz
```
Comment (MarcoFalke) [inline Dockerfile]: Wouldn't it make more sense to only have a minimum set of build dependencies and then build the Bitcoin Core dependencies from `./depends/`?

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    && cd /usr/src \
+    && rm -rf /usr/src/db-${DB4_VERSION}
+
+# Build
+RUN set -x \
+    && cd /usr/src/bitcoin \
+    && ./autogen.sh \
+    && ./configure --with-gui=no \
+    && make -j$(nproc) \
+    && make install \
```
Comment (MarcoFalke) [inline Dockerfile]: If there goal is to provide an easy entry point for development, ci or testing, the `install` step shouldn't be run as part of the image creation, I believe. It should instead be provided though the docker compose entrypoint. See for example the work by @jamesob https://github.com/bitcoin/bitcoin/pull/10903/files

**Comment by MarcoFalke** [comment]
Concept ACK, if this helps people to run tests, ci and generally makes it easier to bootstrap a development environment. Though, we shouldn't provide any ways to run this in production, since there are already statically linked, gitian-built binaries available. If you need those releases wrapped into a dockerfile, there is already https://github.com/NicolasDorier/docker-bitcoin/tree/master/core

**Comment by stakauskas** [comment]
> Concept ACK, if this helps people to run tests, ci and generally makes it easier to bootstrap a development environment.

The initial pull request emerged and divided into two groups dev/prod. Initially it was created as an alternative to running standalone binaries in favor of running it in containerised setup. If it's completely for dev purposes (which is a good start) I'll modify Dockerfile and related files accordingly so it could be more suitable for the purpose.

**Comment by skwp** [comment]
Going to chime in again to slightly disagree with "we shouldn't provide docker for production", again due to the "source of truth" argument. If you google bitcoin on docker you get a variety of blog posts and repos. It's not that hard to provide a dockerfile and make sure it's working which ensures that it's as easy as possible to run a node. If more nodes is what we want, then providing a production friendly way to spin one up is not a bad idea.

**Comment by promag** [comment]
-1 to provide production images here. I think we shouldn't influence deployment.

+1 in providing docker compose services to build and test with diferent toochain and configurations.

+1 in providing official docker images (not sure how to tackle that) elsewhere.

**Comment by jtimon** [comment]
Concept ACK. This is contrib anyway, if it goes unmaintained it can always be simply removed.

**Comment by NicolasDorier** [comment]
@jtimon it can't be unmaintained if included inside CI. Having a docker failing to build is equivalent to having an out of date documentation: the documentation need to be fixed, or the committer need to not break things.

**Comment by NicolasDorier** [comment]
@sipsorcery you might be interested to follow this PR.

**Comment by promag** [comment]
Summoning @ruimarinho as he also maintains https://github.com/ruimarinho/docker-bitcoin-core.

**Comment by MarcoFalke** [comment]
I believe providing a dockerfile for production use can be done independently (in a different pull request) of providing a dockerfile for development/testing purposes?

**Comment by MarcoFalke** [comment]
If this pull request is to be used to add a devel dockerfile, I'd suggest prefixing the title with "test" or "qa".

**Comment by stakauskas** [comment]
> If this pull request is to be used to add a devel dockerfile, I'd suggest prefixing the title with "test" or "qa".

Files in this pull request would match the prod setup. We use it the way it is to run and scale nodes in prod environment. I'd suggest that better way would be to create a separate pull req for development environment as it will differ significantly and close this if there's no need for prod env. Though, I would agree @skwp that official production friendly docker environment would be nice to have. As for someone having resources it's as easy as one command to run x prod/test nodes in the current containerised infrastructure. And I don't see how it would influence deployment as mentioned by @promag apart from the fact that it would make it much easier to deploy and manage.

**Comment by promag** [comment]
I mean, should we promote deployments with docker? The way I see, that would be the case if a dockerfile is added.

**Comment by stakauskas** [comment]
> I mean, should we promote deployments with docker? The way I see, that would be the case if a dockerfile is added.

It wouldn't be a promotion, it would just be an ability for people to choose another variant to running the software. Most of the projects provides official Dockerfiles/docker images in addition to the traditional released binaries. It would also make it easy to have official way to run bitcoin nodes in a container native infrastructures which is quite common approach by todays standards. It would possibly attract much more people running nodes, with having official way to do it.

**Comment by TheBlueMatt** [comment]
We now also have an external "packaging" repo at https://github.com/bitcoin-core/packaging where things like this may make more sense, just to add to the discussion.

**Comment by stakauskas** [comment]
Agreed, moving Docker related stuff to bitcoin-core/packaging would make sense.

**Comment by jtimon** [comment]
> @TheBlueMatt Agreed, moving Docker related stuff to bitcoin-core/packaging would make sense.

That would solve the "6 people maintaining the same thing independently" problem too. Fine with me.
To me the most important part is avoiding that work duplication.

**Comment by MarcoFalke** [comment]
Could you adjust this pull request to be just a patch that adds a link in some doc to https://github.com/bitcoin-core/packaging/blob/master/docker/README.md ?

**Comment by MarcoFalke** [comment]
And everyone who is interested in this should go and review https://github.com/bitcoin-core/packaging/pull/2

**Comment by NicolasDorier** [comment]
I really dislike adding on `bitcoin-core/packaging`... the whole point is having that run as part of the CI. By moving on separate repo, it makes sure it will never be part of CI because you can't build with the specific commit you pulled via `COPY . .`

File: contrib/docker/README.md [hunk 3]
```diff
@@ -0,0 +1,119 @@
+Docker setup
```
Comment (practicalswift) [inline contrib/docker/README.md]: FWIW, various markdown cleanups suggested by `mdl`:

```
job=2018-09-11-16-17-08 pr=13542 mdl: contrib/docker/README.md:1: MD002 First header should be a top level header
job=2018-09-11-16-17-08 pr=13542 mdl: contrib/docker/README.md:4: MD003 Header style
job=2018-09-11-16-17-08 pr=13542 mdl: contrib/docker/README.md:9: MD003 Header style
job=2018-09-11-16-17-08 pr=13542 mdl: contrib/docker/README.md:13: MD003 Header style
job=2018-09-11-16-17-08 pr=13542 mdl: contrib/docker/README.md:17: MD003 Header style
job=2018-09-11-16-17-08 pr=13542 mdl: contrib/docker/README.md:24: MD003 Header style
job=2018-09-11-16-17-08 pr=13542 mdl: contrib/docker/README.md:34: MD003 Header style
job=2018-09-11-16-17-08 pr=13542 mdl: contrib/docker/README.md:38: MD003 Header style
job=2018-09-11-16-17-08 pr=13542 mdl: contrib/docker/README.md:49: MD003 Header style
job=2018-09-11-16-17-08 pr=13542 mdl: contrib/docker/README.md:53: MD003 Header style
job=2018-09-11-16-17-08 pr=13542 mdl: contrib/docker/README.md:57: MD003 Header style
job=2018-09-11-16-17-08 pr=13542 mdl: contrib/docker/README.md:61: MD003 Header style
job=2018-09-11-16-17-08 pr=13542 mdl: contrib/docker/README.md:63: MD003 Header style
job=2018-09-11-16-17-08 pr=13542 mdl: contrib/docker/README.md:73: MD003 Header style
job=2018-09-11-16-17-08 pr=13542 mdl: contrib/docker/README.md:75: MD003 Header style
job=2018-09-11-16-17-08 pr=13542 mdl: contrib/docker/README.md:82: MD003 Header style
job=2018-09-11-16-17-08 pr=13542 mdl: contrib/docker/README.md:95: MD003 Header style
job=2018-09-11-16-17-08 pr=13542 mdl: contrib/docker/README.md:99: MD003 Header style
job=2018-09-11-16-17-08 pr=13542 mdl: contrib/docker/README.md:108: MD003 Header style
job=2018-09-11-16-17-08 pr=13542 mdl: contrib/docker/README.md:112: MD003 Header style
job=2018-09-11-16-17-08 pr=13542 mdl: contrib/docker/README.md:19: MD009 Trailing spaces
job=2018-09-11-16-17-08 pr=13542 mdl: contrib/docker/README.md:115: MD009 Trailing spaces
job=2018-09-11-16-17-08 pr=13542 mdl: contrib/docker/README.md:17: MD022 Headers should be surrounded by blank lines
job=2018-09-11-16-17-08 pr=13542 mdl: contrib/docker/README.md:75: MD022 Headers should be surrounded by blank lines
job=2018-09-11-16-17-08 pr=13542 mdl: contrib/docker/README.md:82: MD022 Headers should be surrounded by blank lines
job=2018-09-11-16-17-08 pr=13542 mdl: contrib/docker/README.md:112: MD022 Headers should be surrounded by blank lines
job=2018-09-11-16-17-08 pr=13542 mdl: contrib/docker/README.md:112: MD024 Multiple headers with the same content
job=2018-09-11-16-17-08 pr=13542 mdl: contrib/docker/README.md:18: MD031 Fenced code blocks should be surrounded by blank lines
job=2018-09-11-16-17-08 pr=13542 mdl: contrib/docker/README.md:76: MD031 Fenced code blocks should be surrounded by blank lines
job=2018-09-11-16-17-08 pr=13542 mdl: contrib/docker/README.md:83: MD031 Fenced code blocks should be surrounded by blank lines
job=2018-09-11-16-17-08 pr=13542 mdl: contrib/docker/README.md:113: MD031 Fenced code blocks should be surrounded by blank lines
```

File: contrib/docker/docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,18 @@
+version: '3.6'
```
Comment (practicalswift) [inline contrib/docker/docker-compose.yml]:
```
2018-09-22 22:07:50 yamllint(pr=13542): contrib/docker/docker-compose.yml:1:1: [warning] missing document start "---" (document-start)
```

**Pull request closed without merging**